### PR TITLE
box: move space triggers to trigger module

### LINKED ISSUE #9127: Assertion failed when before_replace trigger changes tuple on recovery from snapshot
**Reproducer:**
```lua
-- gen_snap.lua
equire('os').execute('rm -rf 000*')
box.cfg{checkpoint_interval = 0}

s = box.schema.create_space('test')
s:create_index('pk')
for i = 1, 10 do
	s:replace{i, i - 1}
end
box.snapshot()
require('os').exit()
```

```lua
-- before_recovery_replace.lua
local function test_before_replace_trig(old, new)
    return new:update{{'+', 2, 1}}
end

local function space_on_replace_trig(old, new)
    if new and new[3] == 'test' then
        box.on_commit(function()
            box.space.test:before_replace(test_before_replace_trig)
        end)
    end
end

local function on_init_trig()
    box.space._space:on_replace(space_on_replace_trig)
end

box.ctl.on_schema_init(on_init_trig)

box.cfg{}
require('os').exit()
```

```
tarantool gen_snap.lua && tarantool before_recovery_replace.lua
```

**Result:**
```
tarantool: ./src/box/memtx_space.c:124: memtx_space_replace_build_next: Assertion `old_tuple == NULL && mode == DUP_INSERT' failed.
```

If `box.snapshot()` from `gen_snap.lua` is removed, everything works fine.

The problem is caused by #9125 - `DUP_INSERT` is replaced with `DUP_REPLACE` and this fair assertion is failed.

### PR DESCRIPTION
**Pull request by drewdzzz**
The patch moves space triggers to trigger registry and adds required types to `func_adapter`. Also, an assertion fail on recovery from snapshot with before_replace triggers is fixed along the way.

Part of https://github.com/tarantool/tarantool/issues/8657
Closes https://github.com/tarantool/tarantool/issues/8859
Closes #9127

### CODE REVIEW

File: src/box/lua/space.cc [hunk 1]
```diff
@@ -57,129 +59,70 @@ extern "C" {
...
+			   "[function | nil], [string])");
 	}
 	lua_getfield(L, 1, "id"); /* Get space id. */
 	uint32_t id = lua_tonumber(L, lua_gettop(L));
-	struct space *space = space_cache_find_xc(id);
 	lua_pop(L, 1);
 
-	return lbox_trigger_reset(L, 3, &space->on_replace,
-				  lbox_push_txn_stmt, NULL);
+	if (!box_is_configured()) {
```
[the remark this answers is not in the document]
Comment (alyapunov) [reply src/box/lua/space.cc:93]: Also this part contradicts with the commit message.

File: src/box/space.c [hunk 2]
```diff
@@ -325,6 +326,123 @@ space_unpin_defaults(struct space *space)
...
+	/* Format strings for names of events associated with space by name. */
+	static const char * const regular_event_by_name_fmt[] = {
+		"box.space.%s.on_replace",
+		"box.space.%s.before_replace",
+	};
+	static const char * const recovery_event_by_name_fmt[] = {
+		"box.space.%s.on_recovery_replace",
+		"box.space.%s.before_recovery_replace",
+	};
+	assert(lengthof(space_events) == lengthof(regular_event_by_id_fmt));
```
Comment (alyapunov) [inline src/box/space.c:364]: I would try to use static assert here.

File: src/box/space.c [hunk 3]
```diff
@@ -325,6 +326,123 @@ space_unpin_defaults(struct space *space)
...
+	/*
+	 * Space events - on_replace and before_replace events. All event names
+	 * must be declared in the same order.
+	 */
+	struct space_event *space_events[] = {
+		&space->on_replace_event,
+		&space->before_replace_event,
+	};
+	/* Format strings for names of events associated with space by id. */
+	static const char * const regular_event_by_id_fmt[] = {
```
Comment (alyapunov) [inline src/box/space.c:347]: I would not use 'static' here.

File: src/box/space.c [hunk 4]
```diff
@@ -325,6 +326,123 @@ space_unpin_defaults(struct space *space)
 	}
 }
 
+/**
+ * Set all the events associated with space: on_replace and before_replace,
+ * bound by name and by id. If argument on_recovery is true, recovery events
+ * are set instead of regular ones.
+ */
+static void
+space_set_events(struct space *space, bool on_recovery)
```
Comment (alyapunov) [inline src/box/space.c:335]: We usually use 'is' or 'has' etc prefixes for boolean flags.

File: src/box/space.c [hunk 5]
```diff
@@ -325,6 +326,123 @@ space_unpin_defaults(struct space *space)
...
+	 * So the snippet chooses between tt_static_buf and malloc depending on
+	 * maximal length of the name of event associated by the space name.
+	 */
+	const size_t event_by_name_fmt_max_len = 40;
+	size_t buf_size = event_by_name_fmt_max_len +
+		strlen(space->def->name) + 1;
+	buf_size = MAX(buf_size, TT_STATIC_BUF_LEN);
+	char *event_name = NULL;
+	if (buf_size > TT_STATIC_BUF_LEN)
+		event_name = xmalloc(buf_size);
```
Comment (alyapunov) [inline src/box/space.c:397]: It should be `xregion_alloc` and `region_truncated`.

File: src/box/lua/func_adapter.c [hunk 6]
```diff
@@ -197,8 +206,8 @@ func_adapter_lua_pop_tuple(struct func_adapter_ctx *base, struct tuple **out)
 }
 
 /**
- * Null in Lua can be represented in three ways: nil, box.NULL or just absence
- * of an object. The function checks all the cases.
+ * Null in Lua can be represented in two ways: nil or box.NULL.
```
[the remark this answers is not in the document]
Comment (drewdzzz) [reply src/box/lua/func_adapter.c]: Fixed.

File: src/box/lua/space.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (drewdzzz) [reply src/box/lua/space.cc]: 1. Use new `space->run_recovery_truggers` flag for this purpose.
2. Updated the commit message.

File: src/box/space.c (hunk 3 above)
Comment (drewdzzz) [reply src/box/space.c]: Cannot use just `const char *` because of checkpatch.

File: src/box/space.c (hunk 4 above)
Comment (drewdzzz) [reply src/box/space.c]: Renamed to `is_on_recovery`.

File: src/box/space.c (hunk 5 above)
Comment (drewdzzz) [reply src/box/space.c]: Done.

**Review by ochaplashkin — APPROVED** [review]

**Review by alyapunov — APPROVED** [review]

**Pull request merged**